README.md
# ewN2HDECAY

A program for the Calculation of Electroweak One-Loop Corrections to Higgs Decays in the Next-to-Minimal Two-Higgs-Doublet Model Including State-of-the-Art QCD Corrections

## Program information

**Program** ewN2HDECAY 1.0.3

**Date** 06.12.2019

**Authors** [Marcel Krause](mailto:marcel.krause@alumni.kit.edu) and [M. Margarete Mühlleitner](mailto:margarete.muehlleitner@kit.edu)

**Manual** [Comp. Phys. Commun. 247 (2020) 106924](https://www.sciencedirect.com/science/article/pii/S001046551930298X), [arXiv:1904.02103 [hep-ph]](https://arxiv.org/abs/1904.02103)

**Citations** When you use this program, please acknowledge the work of our and other groups by citing the following papers:
- The manual for ewN2HDECAY:
  - M. Krause, M. Muhlleitner, Comp. Phys. Commun. 247 (2020) 106924, arXiv:1904.02103
- The manuals for HDECAY and N2HDECAY:
  - A. Djouadi, J. Kalinowski, M. Spira, Comp. Phys. Commun. 108 (1998) 56, hep-ph/9704448
  - A. Djouadi, J. Kalinowski, M. Muhlleitner, M. Spira, arXiv:1801.09506 (hep-ph)
  - M. Muhlleitner, M. O. P. Sampaio, R. Santos, J. Wittbrodt, JHEP 1703 (2017) 094, arXiv:1612.01309 (hep-ph)
- The paper on the electroweak correction to the N2HDM decays:
  - M. Krause, D. Lopez-Val, M. Muhlleitner, R. Santos, JHEP 12 (2017) 077, arXiv:1708.01578 (hep-ph)
- The publication of LoopTools:
  - T. Hahn, M. Perez-Victoria, Comp. Phys. Commun. 118 (1999) 153-165, hep-ph/9807565

**Abstract** We present the program package ewN2HDECAY for the calculation of the partial decay widths and branching ratios of the Higgs bosons of a general CP-conserving Next-to-Minimal 2-Higgs doublet model (N2HDM). The tool includes the full electroweak one-loop corrections to all two-body onshell Higgs decays in the N2HDM that are not loop-induced. It combines them with the state-of-the-art QCD corrections that are already implemented in the program HDECAY. For the renormalization of the electroweak sector an on-shell scheme is implemented for most of the renormalization parameters. Exceptions are the soft-![](https://latex.codecogs.com/gif.latex?%5Cmathbb%7BZ%7D_2 "\mathbb{Z}_2")-breaking squared mass scale ![](https://latex.codecogs.com/gif.latex?m_%7B12%7D%5E2 "m_{12}^2") and the singlet vacuum expectation value ![](https://latex.codecogs.com/gif.latex?v_S "v_S"), where an ![](https://latex.codecogs.com/gif.latex?%5Coverline%7B%5Ctext%7BMS%7D%7D "\overline{\text{MS}}") condition is applied, as well as the N2HDM mixing angles ![](https://latex.codecogs.com/gif.latex?%5Calpha%20_i%20~%28i%3D1%2C2%2C3%29 "\alpha _i~(i=1,2,3)") and ![](https://latex.codecogs.com/gif.latex?%5Cbeta "\beta"), for which several distinct renormalization schemes are implemented. The tool ewN2HDECAY can be used for phenomenological analyses of the branching ratios of Higgs decays in the N2HDM. Furthermore, the separate output of the electroweak contributions to the tree-level partial decay widths for several different renormalization schemes allows for an efficient analysis of the impact of the electroweak corrections and the remaining theoretical error due to missing higher-order corrections. The latest version of the program package ewN2HDECAY can be downloaded from the URL https://github.com/marcel-krause/ewN2HDECAY.

**Changelog** For a documentation about the changes made in ewN2HDECAY, check the [Changelog.md](Changelog.md) file.

**Copyright** Copyright (C) 2019, Marcel Krause and Milada Margarete Mühlleitner

**License** GNU General Public License (GNU GPL-3.0-or-later). ewN2HDECAY is released under GNU General Public License (GNU GPL-3.0-or-later). This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version. This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details. You have received a copy ([LICENSE.md](LICENSE.md)) of the GNU General Public License along with this program.

**Contact** For feedback, complaints and bug reports, please send an e-mail to <marcel.krause@alumni.kit.edu> and <margarete.muehlleitner@kit.edu>.

## Getting Started

For more detailed information about installing and using the program, please have a look at the [full manual](Documentation/manual.pdf "full manual").

### System requirements

Supported operating systems:
- Windows 7 and Windows 10 (with [Cygwin](https://www.cygwin.com/ "Cygwin") installed)
- Linux
- macOS

The following components must be installed on your system to compile and run ewN2HDECAY:
- GNU gcc (tested with versions 6.4.0 and 7.3.1)
- GNU g++
- GNU gfortran
- Python 2 or 3 (tested with versions 2.7.14 and 3.5.0)
- find
- cURL

If you install ewN2HDECAY on Windows, make sure that you install [Cygwin](https://www.cygwin.com/ "Cygwin") first, including the following components:
- GNU gcc
- GNU g++
- GNU gfortran
- find
- cURL

### Installing

For an easy installation, we recommend using the automatic installer of ewN2HDECAY, which guides you through the installation. Download the latest version of ewN2HDECAY from https://github.com/marcel-krause/ewN2HDECAY. Open a shell, navigate to the ewN2HDECAY root folder and execute the following command:
```
python setup.py
```

The installer asks you if you want to download and install LoopTools, which is required for ewN2HDECAY to run. If you do not have LoopTools installed on your system already type y. The installer downloads the LoopTools version specified in the file [Config.py](Config.py); if you prefer another version, change the corresponding entry in that file first before executing the setup.
If you already have LoopTools on your system, you can choose not to download and install LoopTools. In that case, open the file [Config.py](Config.py) and set
```
useRelativeLoopToolsPath = False
```
and change the variable pathLoopTools to the *absolute* path to the LoopTools main folder (containing the bin, include and library subfolders). Additionally, modify the variables pathLoopToolsLibs and pathLoopToolsExecs to the relative paths to the library and bin subfolders with respect to the LoopTools main directory.

The installer further asks you whether the makefile and electroweakCorrections.F90 file shall be created and whether it should make the program. Type y for all these questions in order to compile ewN2HDECAY.

### Using ewN2HDECAY

Using ewN2HDECAY is simple. Save all input files that you want to use in the 'Input' subdirectory. You can put as many input files in the directory as you wish and the input files can have arbitrary file names. Make sure that the input files have the exact input format required by ewN2HDECAY, as described in the [full manual](Documentation/manual.pdf "full manual").

To start ewN2HDECAY, execute the following command in the main folder of ewN2HDECAY:
```
python ewN2HDECAY.py
```
The program will iterate over all input files and calculate all higher-order corrections according to the parameter and renormalization scheme choices as given in your input files. The resulting output files are saved with the same file name as the input files in the 'Results' subdirectory. For each input file, two output files are generated: one file containing the branching ratios with and without the electroweak corrections, indicated by a filename suffix '_BR', as well as one file containing the electroweak partial decay widths at tree-level and one-loop order, indicated by the filename suffix '_EW'.


Changelog.md
## [1.0.3] - 2019-06-12
### Added
- Added the journal information.
### Changed
-  Changed the mail address of one of the authors.

## [1.0.2] - 2019-05-08
### Changed
- Fixed an uncritical typo in the decay of H2 to two gluons.

## [1.0.1] - 2019-05-08
### Changed
- Fixed a bug which prevented the subprogram N2HDECAY from installing.
- Added the makefile for the N2HDECAY subprogram and removed the gitignore statement for this file.

ewN2HDECAY.py
#!/usr/bin/env python
#Filename: ewN2HDECAY.py


 #################################################################################################################################
#                                                                                                                                 #
#                                                                 ewN2HDECAY                                                      #
#                                                                                                                                 #
#   Purpose:    A program for the calculation One-Loop Electroweak Corrections to Higgs Decays in the Next-to-Minial              #
#               Two-Higgs-Doublet Model (N2HDM) Including State-of-the-Art QCD Corrections                                        #
#   Authors:    Dr. Marcel Krause (marcel.krause@alumni.kit.edu)                                                                  #
#               Prof. Dr. M. Margarete Muehlleitner (margarete.muehlleitner@kit.edu)                                              #
#   Version:    1.0.3                                                                                                             #
#   Date:       06.12.2019                                                                                                        #
#   Copyright:  Copyright (C) 2019, Marcel Krause and Milada Margarete Muehlleitner                                               #
#   License:    GNU General Public License (GNU GPL-3.0-or-later)                                                                 #
#                                                                                                                                 #
#               ewN2HDECAY is released under GNU General Public License (GNU GPL-3.0-or-later). This program is free software:    #
#               you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the    #
#               Free Software Foundation, either version 3 of the License, or any later version.                                  #
#                                                                                                                                 #
#               This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the        #
#               implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for   #
#               more details.                                                                                                     #
#                                                                                                                                 #
#               You have received a copy (LICENSE.md) of the GNU General Public License along with this program in the            #
#               ewN2HDECAY root directory.                                                                                        #
#                                                                                                                                 #
#   Citation:   When you use this program, please acknowledge the work of our and other groups by citing the following papers:    #
#                   The manual for ewN2HDECAY:                                                                                    #
#                    - M. Krause, M. Muhlleitner, Comp. Phys. Commun. 247 (2020) 106924, arXiv:1904.02103 (hep-ph)                #
#                   The manuals for HDECAY and N2HDECAY:                                                                          #
#                    - A. Djouadi, J. Kalinowski, M. Spira, Comp. Phys. Commun. 108 (1998) 56, hep-ph/9704448                     #
#                    - A. Djouadi, J. Kalinowski, M. Muhlleitner, M. Spira, arXiv:1801.09506 (hep-ph)                             #
#                    - M. Muhlleitner, M. O. P. Sampaio, R. Santos, J. Wittbrodt, JHEP 1703 (2017) 094, arXiv:1612.01309 (hep-ph) #
#                   The paper on the EW correction to the N2HDM decays:                                                           #
#                    - M. Krause, D. Lopez-Val, M. Muhlleitner, R. Santos, JHEP 12 (2017) 077, arXiv:1708.01578 (hep-ph)          #
#                   The publication of LoopTools:                                                                                 #
#                    - T. Hahn, M. Perez-Victoria, Comp. Phys. Commun. 118 (1999) 153-165, hep-ph/9807565                         #
#                                                                                                                                 #
 #################################################################################################################################


#------------------------------#
#         Import Modules       #
#------------------------------#
import sys
import os
from shutil import copyfile, rmtree
from math import pi, sqrt
import subprocess
# import multiprocessing
import CommonFunctions      # Provides common, often used functions for different scripts of ewN2HDECAY

#-------------------------#
#        Settings         #
#-------------------------#
# WARNING: do not change these settings if you do not know what they do!
lineToInsert = 134      # This is the line at which the temporary input file ends and at which we append the electroweak corrections
lineWhereAlphaAtMZ = 27 # This is the line at which in the temporary input file the fine-structure constant at the Z boson mass MZ is specified
lineWhereGFCalc = 29    # This is the line at which in the temporary input file the calculated Fermi constant GFCALC is specified
lineWhereMZ = 32        # This is the line at which in the temporary input file the Z boson mass MZ is specified
lineWhereMW = 33        # This is the line at which in the temporary input file the W boson mass MW is specified
lineWhereOSMC = 23      # This is the line at which the OS MC value has to be inserted in the temporary input file
lineWhereOSMB = 24      # This is the line at which the OS MB value has to be inserted in the temporary input file
lineWhereParamType = 57 # This is the line at which the parameter type is specified
lineWhereRefScheme = 77 # This is the line at which the reference renormalization scheme is specified

#----------------------------#
#        Main Program        #
#----------------------------#

if __name__ == "__main__":
	# Print the welcome screen
	print('''
+---------------------------------------+
|                                       |
|           ewN2HDECAY 1.0.3            |
|                                       |
|                             /         |
|                            /          |
|                           /           |
|                      --- /            |
|      ______________/     \            |
|                    \     /            |
|                      --- \            |
|                           \           |
|                            \          |
|                             \         |
|                                       |
+---------------------------------------+

When you use this program please cite:
	The manual for ewN2HDECAY:
	 - M. Krause, M. Muhlleitner, Comp. Phys. Commun. 247 (2020) 106924, arXiv:1904.02103 (hep-ph)
	The manuals for HDECAY and N2HDECAY:
	 - A. Djouadi, J. Kalinowski, M. Spira, Comp. Phys. Commun. 108 (1998) 56, hep-ph/9704448
	 - A. Djouadi, J. Kalinowski, M. Muhlleitner, M. Spira, arXiv:1801.09506 (hep-ph)
	 - M. Muhlleitner, M. O. P. Sampaio, R. Santos, J. Wittbrodt, JHEP 1703 (2017) 094, arXiv:1612.01309 (hep-ph)
	The paper on the EW correction to the N2HDM decays:
	 - M. Krause, D. Lopez-Val, M. Muhlleitner, R. Santos, JHEP 12 (2017) 077, arXiv:1708.01578 (hep-ph)
	The publication of LoopTools:
	 - T. Hahn, M. Perez-Victoria, Comp. Phys. Commun. 118 (1999) 153-165, hep-ph/9807565

ewN2HDECAY is released under GNU General Public License (GNU GPL-3.0-or-later). This program is free software: 
you can redistribute it and/or modify it under the terms of the GNU General Public License as published by 
the Free Software Foundation, either version 3 of the License, or any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the 
implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
for more details.

You have received a copy (LICENSE.md) of the GNU General Public License along with this program in the ewN2HDECAY
root directory.

Copyright 2019, Marcel Krause and Milada Margarete Muehlleitner.
	''')

	# Get a list of all input files
	inputPath = "Input"
	inputFileList = os.listdir(inputPath)
	if '..' in inputFileList:
		inputFileList.remove('..')
	if '.' in inputFileList:
		inputFileList.remove('.')

	# Iterate over all input files
	for inputFileTemp in inputFileList:
		print("Calculating corrections for input file " + inputFileTemp + " ...\n")

		# Copy the input file to the N2HDECAY subfolder
		print("Copying input files into N2HDECAY folder...")
		filenameIn = "Input" + os.sep + inputFileTemp
		filenameOut = "N2HDECAY" + os.sep + "n2hdecay.in"
		# Remove any existing input and fermion masses file in N2HDECAY
		if os.path.isfile(filenameOut):
			os.remove(filenameOut)
		if os.path.isfile("N2HDECAY" + os.sep + "fermionmasses.dat"):
			os.remove("N2HDECAY" + os.sep + "fermionmasses.dat")
		if os.path.isfile("N2HDECAY" + os.sep + "alphaandbeta.dat"):
			os.remove("N2HDECAY" + os.sep + "alphaandbeta.dat")
		copyfile(filenameIn, filenameOut)
		with open("N2HDECAY" + os.sep + "alphaandbeta.dat", 'w') as fileHandler:
			fileHandler.write(' alpha1 = 0D0\n alpha2 = 0D0\n alpha3 = 0D0\n beta  = 0D0\n')
		print("... done.\n")

		# Let N2HDECAY run in minimal mode to produce the fermion mass file
		print("Starting N2HDECAY in minimal mode...")
		os.chdir('N2HDECAY')
		prompt = ['./run', '1']
		subprocess.call(prompt, stdin=None, stdout=None, stderr=None, shell=False)
		os.chdir('..')
		print("N2HDECAY in minimal mode terminated.\n")

		# Read the fermion masses from the fermion mass file
		filenameMasses = "N2HDECAY" + os.sep + "fermionmasses.dat"
		massFileLines = list(line.rstrip('\n') for line in open(filenameMasses))
		MCOSCalc = float((massFileLines[0].split('='))[1].strip())
		MBOSCalc = float((massFileLines[1].split('='))[1].strip())
		
		# Copy the file name to the N2HDECAY folder and truncate it at the end
		print("Copying input files into N2HDECAY folder...")
		filenameIn = "Input" + os.sep + inputFileTemp
		filenameOut = "N2HDECAY" + os.sep + "n2hdecay.in"
		if os.path.isfile(filenameOut):
			os.remove(filenameOut)
		fileHandler = open(filenameIn, "r")
		convertedFileHandler = []
		lineCount = 1
		for line in fileHandler:
			# If the parameter type is 2, then the reference renormalization scheme has to be set to zero
			if (lineCount == lineWhereRefScheme):
				refScheme = int((line.split())[2])
				if (refScheme == 0):
					print("\nERROR: REFSCHEM=0 is given as input, but REFSCHEM must be greater than zero.")
					print("ewN2HDECAY will be terminated now (ERROR: REFSCHEM=0 was set).")
					sys.exit()
				if (refScheme != 0) and (paramType == 2):
					convertedFileHandler.append("REFSCHEM = 0\n")
					print("\nWARNING: REFSCHEM (line 59) is given as a non-zero value, but TYPE=2 (line 57, lambdas as input) is set.")
					print("REFSCHEM is overwritten to zero, automatic parameter conversion is deactivated!")
					lineCount += 1
					continue
			# Write the current line in an array
			convertedFileHandler.append(line)
			# Check for the parameter type
			if lineCount == lineWhereParamType:
				paramType = int((line.split())[2])
			# Pick out the values of MW, MZ and alphaAtMZ
			if lineCount == lineWhereMZ:
				massMZ = float((line.split())[2])
			if lineCount == lineWhereMW:
				massMW = float((line.split())[2])
			if lineCount == lineWhereAlphaAtMZ:
				alphaAtMZ = float((line.split())[2])
			lineCount += 1
		fileHandler.close()

		# Write a copy of the file to the output folder, but add GFCALC, MCOSCALC and MBOSCALC with the calculated values
		GFcalc = pi/sqrt(2)*alphaAtMZ/(massMW**2*(1-massMW**2/massMZ**2))
		GFline = "GFCALC   = " + str(GFcalc) + "\n"
		MCOSline = "MCOSCALC = " + str(MCOSCalc) + "\n"
		MBOSline = "MBOSCALC = " + str(MBOSCalc) + "\n"
		lineCount = 1
		convertedFile = ''
		renScaleIsDynamic = '0'
		for line in convertedFileHandler:
			if "OUTSCALE" in line and "MIN" in line:
				renScaleIsDynamic = '1'
			if lineCount == lineWhereGFCalc:
				convertedFile += GFline
			elif lineCount == lineWhereOSMC:
				convertedFile += MCOSline
				convertedFile += MBOSline
				convertedFile += line
			else:
				convertedFile += line
			lineCount += 1
		fileHandler = open(filenameOut, "w")
		fileHandler.write(convertedFile)
		fileHandler.close()
		print("... done.\n")
		
		# Calculate the electroweak corrections
		print("Calculating electroweak corrections...\n")
		prompt = ['./electroweakCorrections', '0', '0', '0', '1', 'N2HDECAY' + os.sep + 'n2hdecay.in', 'n2hdecay.in', renScaleIsDynamic]
		subprocess.call(prompt, stdin=None, stdout=None, stderr=None, shell=False)
		print("Calculation of electroweak corrections done.\n")

		# Replace the newline character in each file with a proper newline
		print("Postprocessing temporary input file...")
		fileHandler = open(filenameOut, "r")
		convertedFile = ''
		lineCount = 1
		for line in fileHandler:
			# Convert the literal newlines to actual ones and remove the leading whitespace from the Fortran output
			if (lineCount == lineToInsert):
				lineToReplace = line.replace('\\n', '\n')[1:]
			else:
				lineToReplace = line.replace('\\n', '\n')
			convertedFile += lineToReplace
			# print(lineToReplace)
			lineCount += 1
		fileHandler.close()

		# Store the results file in the correct directory
		if os.path.isfile(filenameOut):
			os.remove(filenameOut)
		fileHandler = open(filenameOut, "w")
		fileHandler.write(convertedFile)
		fileHandler.close()
		print("... done.\n")

		# Start N2HDECAY in the normal (non-minimal) configuration
		print("Starting N2HDECAY in standard mode...")
		os.chdir('N2HDECAY')
		prompt = ['./run']
		subprocess.call(prompt, stdin=None, stdout=None, stderr=None, shell=False)
		os.chdir('..')
		print("N2HDECAY in standard mode terminated.\n")

		# Copy the output files to the results folder
		print("Copying input files into output folder...")
		filenameIn = "N2HDECAY" + os.sep + "slha.out"
		filenameOut = "Results" + os.sep + inputFileTemp.replace('.in', '_BR.out')
		copyfile(filenameIn, filenameOut)
		filenameIn = "N2HDECAY" + os.sep + "ewpartialwidth.out"
		filenameOut = "Results" + os.sep + inputFileTemp.replace('.in', '_EW.out')
		copyfile(filenameIn, filenameOut)
		print("... done.\n")

		# Cleaning
		if os.path.isfile("N2HDECAY" + os.sep + "fermionmasses.dat"):
			os.remove("N2HDECAY" + os.sep + "fermionmasses.dat")
		if os.path.isfile("N2HDECAY" + os.sep + "alphaandbeta.dat"):
			os.remove("N2HDECAY" + os.sep + "alphaandbeta.dat")

		print("Corrections for input file " + inputFileTemp + " done.\n")

	# End of program is reached
	print("All calculations finished. Thanks for using ewN2HDECAY!")

	# Print the end screen 
	print('''
When you use this program please cite:
	The manual for ewN2HDECAY:
	 - M. Krause, M. Muhlleitner, Comp. Phys. Commun. 247 (2020) 106924, arXiv:1904.02103 (hep-ph)
	The manuals for HDECAY and N2HDECAY:
	 - A. Djouadi, J. Kalinowski, M. Spira, Comp. Phys. Commun. 108 (1998) 56, hep-ph/9704448
	 - A. Djouadi, J. Kalinowski, M. Muhlleitner, M. Spira, arXiv:1801.09506 (hep-ph)
	 - M. Muhlleitner, M. O. P. Sampaio, R. Santos, J. Wittbrodt, JHEP 1703 (2017) 094, arXiv:1612.01309 (hep-ph)
	The paper on the EW correction to the N2HDM decays:
	 - M. Krause, D. Lopez-Val, M. Muhlleitner, R. Santos, JHEP 12 (2017) 077, arXiv:1708.01578 (hep-ph)
	The publication of LoopTools:
	 - T. Hahn, M. Perez-Victoria, Comp. Phys. Commun. 118 (1999) 153-165, hep-ph/9807565

ewN2HDECAY is released under GNU General Public License (GNU GPL-3.0-or-later). This program is free software: 
you can redistribute it and/or modify it under the terms of the GNU General Public License as published by 
the Free Software Foundation, either version 3 of the License, or any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the 
implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
for more details.

You have received a copy (LICENSE.md) of the GNU General Public License along with this program in the ewN2HDECAY
root directory.

Copyright 2019, Marcel Krause and Milada Margarete Muehlleitner.
	''')

	sys.exit()
